### PR TITLE
snap: fix SNAP* environment merging in `snap run`

### PR DESCRIPTION
**Pull request by mvo5**

### CODE REVIEW

File: cmd/snap/cmd_run.go [hunk 1]
```diff
@@ -193,6 +193,29 @@ func snapRunHook(snapName, snapRevision, hookName string) error {
 	return runSnapConfine(info, hook.SecurityTag(), snapName, "", hook.Name, nil)
 }
 
+// FIXME: move to osutil?
+func envMap(env []string) map[string]string {
+	envMap := map[string]string{}
```
Comment (zyga) [inline cmd/snap/cmd_run.go]: did you intend to `make(map[string]string{}` here? (not sure how go behaves ATM)

File: cmd/snap/cmd_run.go [hunk 2]
```diff
@@ -217,7 +240,11 @@ func runSnapConfine(info *snap.Info, securityTag, snapApp, command, hook string,
 	cmd = append(cmd, snapApp)
 	cmd = append(cmd, args...)
 
-	env := append(os.Environ(), snapExecEnv(info)...)
+	env := envMap(os.Environ())
+	snapEnv := envMap(snapExecEnv(info))
+	for k, v := range snapEnv {
```
Comment (zyga) [inline cmd/snap/cmd_run.go]: I'd add a comment that this explicitly overrides all of the values from `snapExecEnv(...)`

**Comment by zyga** [comment]
LGTM

**Comment by chipaca** [comment]
given that only `cmd_run` uses `snapenv`, shouldn't we fix things there instead of adding business logic to snap?

**Comment by chipaca** [comment]
Perhaps something like
```go
func ExecEnv(info *snap.Info) {
	home := os.Getenv("HOME")
	// HOME is not set for systemd services, so pull it out of passwd
	if home == "" {
		user, err := user.Current()
		if err == nil {
			home = user.HomeDir
		}
	}

	m := Basic(info)
	if home != "" {
		addUserEnv(m, info, home)
	}

	osEnviron := os.Environ()
	env := make([]string, 0, len(m)+len(osEnviron))
	for k, v := range m {
		env = append(env, fmt.Sprintf("%s=%s", k, v))
	}
	for _, v := range osEnviron {
		idx := strings.IndexByte(v, '=')
		if idx <= 0 {
			continue
		}
		k := v[:idx]
		if _, isSnapEnv := m[k]; !isSnapEnv {
			env = append(env, v)
		}
	}

	return env
}
```

**Comment by mvo5** [comment]
@chipaca Thank you! I pushed a revision that moves it all into snapenv and unexports some bits from there.

File: snap/snapenv/snapenv.go [hunk 3]
```diff
@@ -22,16 +22,31 @@ package snapenv
...
+	}
+	return envFromMap(env)
+}
+
+// basicEnv returns the app-level environment variables for a snap.
 // Despite this being a bit snap-specific, this is in helpers.go because it's
 // used by so many other modules, we run into circular dependencies if it's
 // somewhere more reasonable like the snappy module.
-func Basic(info *snap.Info) []string {
+func basicEnv(info *snap.Info) []string {
```
Comment (chipaca) [inline snap/snapenv/snapenv.go]: note that here you're building a `[]string` and then parsing it out into a map, which seems silly (hence why my suggestion (in code, rather obscure, yadda) of making Basic just return a map (and making User take the map and add things to it)).

File: snap/snapenv/snapenv.go [hunk 4]
```diff
@@ -22,37 +22,99 @@ package snapenv
...
+	env := envMap(os.Environ())
+	snapEnv := snapEnv(info)
+	for k, v := range snapEnv {
+		env[k] = v
+	}
+	return envFromMap(env)
+}
+
+// returns the environment that is important for the later stages of execution
+// (like SNAP_REVISION that snap-exec requires to work)
```
Comment (kyrofa) [inline snap/snapenv/snapenv.go]: This comment is likely less relevant in this context.

**Comment by pedronis** [comment]
lgtm

**Comment by kyrofa** [comment]
One note on the comment that was copied over, but looks good to me :+1: .

**Pull request merged**